internal/extensions/resolve.go

```diff
@@ -0,0 +1,167 @@
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+const (
+	legacyExtensionIndex protowire.Number = 51071 // protovalidate versions < v0.2.0
+)
+
+//nolint:gochecknoglobals // static data, only want single instance
+var resolver = newExtensionResolver()
+
+// Resolve resolves extensions without using [proto.GetExtension], in case the
+// underlying type of the extension is not the concrete type expected by the
+// library. In some cases, particularly when using a dynamic descriptor set, the
+// underlying extension value's type will be a dynamicpb.Message. In some cases,
+// the extension may not be resolved at all. This function handles reparsing the
+// fields as needed to get it into the right concrete message as needed. Resolve
+// does not modify the input protobuf message, so it can be used concurrently.
+func Resolve[C proto.Message](
+	options proto.Message,
+	extensionType protoreflect.ExtensionType,
+) (typedMessage C) {
+	message := resolver.resolve(options, extensionType)
+	if message == nil {
+		return typedMessage
+	} else if typedMessage, ok := message.(C); ok {
+		return typedMessage
+	}
+	typedMessage, _ = typedMessage.ProtoReflect().New().Interface().(C)
+	b, _ := proto.Marshal(message)
+	_ = proto.Unmarshal(b, typedMessage)
+	return typedMessage
+}
+
+// extensionResolver implements most of the logic of resolving protovalidate
+// extensions.
+type extensionResolver struct {
+	// types is a types that just contains the protovalidate extensions.
+	types *protoregistry.Types
+
+	// legacyExtensionMap is a mapping from current protovalidate extensions to
+	// legacy protovalidate extensions, used for backwards compatibility. This
+	// map will not be modified, so it is safe to read concurrently.
+	legacyExtensionMap map[protoreflect.ExtensionType]protoreflect.ExtensionType
+}
+
+// newExtensionResolver creates a new extension resolver. This is only called at
+// init and will panic if it fails.
+func newExtensionResolver() extensionResolver {
+	resolver := extensionResolver{
+		types:              &protoregistry.Types{},
+		legacyExtensionMap: make(map[protoreflect.ExtensionType]protoreflect.ExtensionType),
+	}
+	resolver.register(validate.E_Field)
+	resolver.register(validate.E_Message)
+	resolver.register(validate.E_Oneof)
+	resolver.register(validate.E_Predefined)
+	resolver.registerLegacy(validate.E_Field)
+	resolver.registerLegacy(validate.E_Message)
+	resolver.registerLegacy(validate.E_Oneof)
+	return resolver
+}
+
+// register registers an extension into the resolver's registry. This is only
+// called at init and will panic if it fails.
+func (resolver extensionResolver) register(extension protoreflect.ExtensionType) {
+	if err := resolver.types.RegisterExtension(extension); err != nil {
+		//nolint:forbidigo // this needs to be a fatal at init
+		panic(err)
```

Comment:
Sorry I didn't realize this comment wasn't submitted.

Reply:
I always avoid `panic` where possible, but protovalidate will just be non-functional if these registrations fail. We want this ti fail at init just like broken gencode would (because if this happens we likely have a gencode issue.)